### PR TITLE
Assignment6 lambda expressions

### PR DESCRIPTION
**Pull request by KorbinWeiler**
Worked with @BillMillerCoding

### CODE REVIEW

File: Calculate/Program.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+public class Program
+{
+    public Func<string?> ReadLine { get; init; }
+    public Action<string> WriteLine { get; init; }
+
+    public Program()
+    {
+        ReadLine = () => { return Console.ReadLine(); };
+        WriteLine = (value) => { Console.WriteLine(value);};
```
Comment (Al3xramirez) [inline Calculate/Program.cs]: Programs constructor should be empty per assignment instructions I believe

File: CalculateTests/ProgramTests.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace Calculate;
+
+public class ProgramTests
+{
+    [Fact]
+    public void ReadLine_NonDefaultValueSet_ReturnsExpectedValue()
+    {
+        Program program = new Program(() => { return "Test Value"; }, (value) => { });
+        Assert.Equal("Test Value", program.ReadLine());
```
Comment (Al3xramirez) [inline CalculateTests/ProgramTests.cs]: Make sure to have the generic version of Assert.Equals -> `Assert.Equals<string>`, You did it in the calculatorTest.cs 👌

**Review by Al3xramirez — COMMENTED** [review]
Create a Console project called "Calculate.". ✔
Define a Program Class✔
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ❌ (Must have empty default constructor)
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔

Fundamentals
Place all shared project properties into a Directory.Build.props file.✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌(Not all Asserts use this)
All of the above should be unit tested ❌ Should add more code coverage
Choose simplicity over complexity ✔

File: Calculate/Program.cs (hunk 1 above)
Comment (KorbinWeiler) [reply Calculate/Program.cs]: You're right, I forgot that requirement while writing this. I would like to point out that this is functionally the same as making them the default value on the property.

File: CalculateTests/CalculatorTests.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
+﻿using System.Reflection;
+
+namespace Calculate;
+
```
Comment (steeley21) [inline CalculateTests/CalculatorTests.cs:4]: You've got good tests here covering your TryCalculate method, but these don't directly validate the operator methods. I'd suggest adding direct unit tests on Add, Subtract, Multiply, and Divide to catch their edge cases and get better coverage on them.

**Review by steeley21 — COMMENTED** [review]
Instructions

Create a Console project called "Calculate". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
Define a Calculator class ✔
Define static Add, Subtract, Multiply, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiply, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔


Fundamentals

Place all shared project properties into a Directory.Build.props file.
Place all shared project items into a Directory.Build.targets file. (optional)
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ❌
Choose simplicity over complexity ✔


Good work! I think you're just missing some unit tests on your operator functions. :)

**Comment by kellanburns** [comment]
- Create a *Console* project called "Calculate". ✔
- Define a Program Class
  - Define two init-only setter properties, `WriteLine` and `ReadLine`, that contain delegates for writing a line of text and reading a line of text respectively ✔
  - Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
  - Set the default behavior for the `WriteLine` and `ReadLine` properties to invoke `System.Console` versions of the methods and add an empty default constructor. ✔
- Define a Calculator class ✔
  - Define static `Add`, `Subtract`, `Multiply`, and `Divide` methods that have two parameters and return a third parameter. ✔
  - Define a read-only property, `MathematicalOperations`, of type `System.Collections.Generics.IReadOnlyDictionary<TKey,TValue>` that:
    - is initialized to a `System.Collections.Generics.Dictionary<<TKey,TValue>` instance that. ✔
      - Uses `char` for the key corresponding to the operators +, -, *, and /. ✔
      - Has values that correspond with the Add, Subtract, Multiply, and Divide methods. ✔
  - Implement a `TryCalculate` method following "TryParse" pattern ✔
    - Valid `calculation` expressions include such strings as "3 + 4", "42 - 2", etc. ✔
    - If there is no whitespace around the operator, you can assume the `calculation` is invalid and return false. Similarly if the operands are not integers. ✔
    - Use `string.Split()`, pattern matching, logical and operators to parse the string in their entirety ✔
    - Index into the `MathematicalOperations` method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
- Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
- Be sure to use the `WriteLine`/`ReadLine` properties on `Program` for testing the input and output of your program. ✔

## Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

- Refactor the redirect portion of the `Program` class into 'ProgramBase` ❌
- Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects ❌
- Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0) ❌

## Fundamentals

- Place all shared project properties into a `Directory.Build.props` file.
- Place all shared project items into a `Directory.Build.targets` file. (optional)
- nullable reference types is enabled  ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional)   ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ✔
- All of the above should be unit tested ❌
- Choose simplicity over complexity ✔

Code looks good, I'd just unit test the operands more. For example, I'd test for the `DivideByZeroException` in the divide method, and I'd refactor your `Main` to call a function that handles the user-input-loop and allows you to test it.

File: CalculateTests/CalculatorTests.cs
Comment (c-stanton) [inline CalculateTests/CalculatorTests.cs:1]: You could add an additional test for dividing by zero in here.

**Comment by c-stanton** [comment]
•	Create a Console project called "Calculate.". ✔
•	Define a Program Class 
o	Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
o	Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
o	Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
•	Define a Calculator class ✔ 
o	Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
o	Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that: 
	is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔ 
	Uses char for the key corresponding to the operators +, -, *, and /. ✔
	Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
o	Implement a TryCalculate method following "TryParse" pattern ✔ 
	Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
	If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
	Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
	Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
•	Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
•	Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔
Fundamentals
•	Place all shared project properties into a Directory.Build.props file.
•	Place all shared project items into a Directory.Build.targets file. (optional)
•	nullable reference types is enabled ✔
•	Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
•	Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
•	and enabled .NET analyzers for both projects ✔
•	For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
•	All of the above should be unit tested ❌(not all cases covered across Program and Calculator)
•	Choose simplicity over complexity ✔

File: CalculateTests/CalculatorTests.cs
Comment (BillMillerCoding) [reply CalculateTests/CalculatorTests.cs:1]: Added :)

**Review by Joshua-Lester3 — COMMENTED** [review]
## Instructions

- Create a _Console_ project called "Calculate". ✔
- Define a Program Class
  - Define two init-only setter properties, `WriteLine` and `ReadLine`, that contain delegates for writing a line of text and reading a line of text respectively ✔
  - Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
  - Set the default behavior for the `WriteLine` and `ReadLine` properties to invoke `System.Console` versions of the methods and add an empty default constructor. ✔
- Define a Calculator class ✔
  - Define static `Add`, `Subtract`, `Multiply`, and `Divide` methods that have two parameters and return a third parameter.  ❌ nit: doesn't return third param
  - Define a read-only property, `MathematicalOperations`, of type `System.Collections.Generics.IReadOnlyDictionary<TKey,TValue>` that:
    - is initialized to a `System.Collections.Generics.Dictionary<<TKey,TValue>` instance that. ✔
      - Uses `char` for the key corresponding to the operators +, -, \*, and /. ✔
      - Has values that correspond with the Add, Subtract, Multiply, and Divide methods. ✔
  - Implement a `TryCalculate` method following "TryParse" pattern ✔
    - Valid `calculation` expressions include such strings as "3 + 4", "42 - 2", etc. ✔
    - If there is no whitespace around the operator, you can assume the `calculation` is invalid and return false. Similarly if the operands are not integers. ✔
    - Use `string.Split()`, pattern matching, logical and operators to parse the string in their entirety ✔
    - Index into the `MathematicalOperations` method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
- Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
- Be sure to use the `WriteLine`/`ReadLine` properties on `Program` for testing the input and output of your program. ❌ Not tested.

## Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

- Refactor the redirect portion of the `Program` class into 'ProgramBase` ❌
  - Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects
- Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0) ❌

## Fundamentals

- Place all shared project properties into a `Directory.Build.props` file.
- Place all shared project items into a `Directory.Build.targets` file. (optional)
- nullable reference types is enabled ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version) ✔
- All of the above should be unit tested ❌ Doesn't seem to be all unit tested. Code coverage report alerts this, but it looks like most of the testing needs to be done in Program.Main
- Choose simplicity over complexity ✔

File: Calculate/Program.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
+﻿using System.Linq.Expressions;
+using System.Runtime.CompilerServices;
+using Microsoft.VisualBasic;
```
Comment (quattro004) [inline Calculate/Program.cs:3]: Is this (Microsoft.VisualBasic) being used?

File: Calculate/Program.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    public Program()
+    {
+    }
+
+    public Program(Func<string> readLine, Action<string> writeLine)
+    {
+        ReadLine = readLine;
+        WriteLine = writeLine;
+    }
```
Comment (quattro004) [inline Calculate/Program.cs:20]: I'm curious about the intention here as the instructions mention "Set the default behavior for the WriteLine and ReadLine properties to invoke **System.Console** versions of the methods". This makes me think I could use something other than the system defined versions.

File: Calculate/Program.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    static void Main()
+    {
+        Program program = new();
+        Calculator calculator = new();
+
+        program.WriteLine("Enter a calculation (e.g.,3 + 4), or blank to quit");
+
+        string? input;
+        while (!string.IsNullOrWhiteSpace(input = program.ReadLine()))
```
Comment (quattro004) [inline Calculate/Program.cs:30]: Consider that having an infinite loop creates a certain kind of interactive experience like a [REPL](https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop) vs if you did not have the loop the app would exit on each invocation and could be used like in an automated tool call fashion.

File: Calculate/Calculator.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    public static double Divide(int a, int b)
+    {
+        if (b == 0)
+        {
+            throw new DivideByZeroException("Denominator cannot be zero.");
+        }
+        return (double)a / b;
+    }
+
+    public bool TryCalculate(string? expression, out double answer)
```
Comment (quattro004) [inline Calculate/Calculator.cs:45]: The instructions mention to "Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it."

File: CalculateTests/CalculatorTests.cs [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+        //Act
+        bool result = calc.TryCalculate("3 + 4", out double answer);
+
+        //Assert
+        Assert.True(result);
+        Assert.Equal<double>(7.0, answer);
+    }
+
+    [Fact]
+    public void TryCalculate_InvalidInput_ReturnsFalse()
```
Comment (quattro004) [inline CalculateTests/CalculatorTests.cs:22]: I wonder if a Theory with inline data would work here? nit: Typically you want only a single assertion for each unit test.

**Review by quattro004 — APPROVED** [review]
Keep up the good work!
